security/advancedtls/advancedtls.go

```diff
@@ -0,0 +1,375 @@
+	MutualAuth bool
+}
+
+func (o *ClientOptions) config() (*tls.Config, error) {
+	config := &tls.Config{
+		ServerName:           o.ServerNameOverride,
+		Certificates:         o.Certificates,
+		GetClientCertificate: o.GetClientCertificate,
+		RootCAs:              o.RootCACerts,
+		InsecureSkipVerify:   true,
+	}
+	return config, nil
+}
+
+func (o *ServerOptions) config() (*tls.Config, error) {
+	if o.Certificates == nil && o.GetCertificate == nil {
+		return nil, fmt.Errorf("either Certificates or GetCertificate must be specified")
+	}
+	if !o.MutualAuth {
+		return &tls.Config{
+			ClientAuth:     tls.NoClientCert,
+			Certificates:   o.Certificates,
+			GetCertificate: o.GetCertificate,
+		}, nil
+	}
+	if o.GetRootCAs == nil && o.RootCACerts == nil {
+		return nil, fmt.Errorf("server needs to provide root CA certs if using mutual TLS")
+	}
+	// We fall back to normal config settings if users don't need to reload root certificates.
+	if o.RootCACerts != nil {
+		return &tls.Config{
+			ClientAuth:     tls.RequireAndVerifyClientCert,
+			Certificates:   o.Certificates,
+			GetCertificate: o.GetCertificate,
+			ClientCAs:      o.RootCACerts,
+		}, nil
+	}
+	return &tls.Config{
+		ClientAuth:     tls.RequireAnyClientCert,
+		Certificates:   o.Certificates,
+		GetCertificate: o.GetCertificate,
+	}, nil
+
+}
+
+// advancedTLSCreds is the credentials required for authenticating a connection using TLS.
+type advancedTLSCreds struct {
+	config     *tls.Config
+	verifyFunc CustomVerificationFunc
+	getRootCAs func(rawConn net.Conn, rawCerts [][]byte) (*x509.CertPool, error)
+	isClient   bool
+}
+
+func (c advancedTLSCreds) Info() credentials.ProtocolInfo {
+	return credentials.ProtocolInfo{
+		SecurityProtocol: "tls",
+		SecurityVersion:  "1.2",
+		ServerName:       c.config.ServerName,
+	}
+}
+
+func (c *advancedTLSCreds) ClientHandshake(ctx context.Context, authority string,
+	rawConn net.Conn) (_ net.Conn, _ credentials.AuthInfo, err error) {
+	// use local cfg to avoid clobbering ServerName if using multiple endpoints
+	cfg := cloneTLSConfig(c.config)
+	// we return the full authority name to users if ServerName is empty without
+	// stripping the trailing port
+	if cfg.ServerName == "" {
+		cfg.ServerName = authority
+	}
+	cfg.VerifyPeerCertificate = buildVerifyFunc(c, cfg.ServerName, rawConn)
+	conn := tls.Client(rawConn, cfg)
+	errChannel := make(chan error, 1)
+	go func() {
+		errChannel <- conn.Handshake()
+	}()
+	select {
+	case err := <-errChannel:
+		if err != nil {
+			return nil, nil, err
+		}
+	case <-ctx.Done():
+		return nil, nil, ctx.Err()
+	}
+	return WrapSyscallConn(rawConn, conn), credentials.TLSInfo{conn.ConnectionState()}, nil
+}
+
+func (c *advancedTLSCreds) ServerHandshake(rawConn net.Conn) (net.Conn,
+	credentials.AuthInfo, error) {
+	cfg := cloneTLSConfig(c.config)
+	// We build server side verification function only when root cert reloading is needed.
+	if c.getRootCAs != nil {
```

Comment:
At first I thought that using `MutualAuth` here would be a more clear indication that we need to set the verifier since `MutualAuth` means that the server will get a client cert that it needs to verify. But looking at `ServerOptions.config`, it seems that `MutualAuth` doesn't always mean verify, it might mean `RequireAnyClientCert` only. Can you please add a comment to clarify what `c.getRootCAs != nil` means here?

Reply:
Yes, the behaviors are:

if server want client cert(mTLS): 
        if c.getRootCAs != nil:
                clientAuth = RequireAnyClientCert, because we have to disable the default check, 
                as mentioned in last comment
        else:
                clientAuth = RequireAndVerify, because basically user didn't specify anything 
                special for server side, and thus we fall back to normal check(also includes 
                hostname check)
        else:
                clientAuth = NoCert

Reply:
I think it's still worth adding a comment about this in the code. It doesn't have to be here, it can be in the config function, basically combined with the comment I think we should add in that function.